doc/arc.1.in
.\" -*- nroff -*-
.\"  This file is part of the arc package
.\"  Copyright (C) 2002, 2003, 2009 by Gregor Klinke
.\"  
.\"  This library is free software: you can redistribute it and/or modify
.\"  it under the terms of the GNU Lesser General Public License as published
.\"  by the Free Software Foundation, either version 3 of the License, or
.\"  (at your option) any later version.
.\"  
.\"  This library is distributed in the hope that it will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU Lesser General Public License for more details.
.\"  
.\"  You should have received a copy of the GNU General Public License
.\"  along with this program.  If not, see <http://www.gnu.org/licenses/>.

.ds VE arc-@arc-version@
.TH arc 1 \*(VE "arc"
.ig
2002/08/06
..
.\" Please update the above date whenever this man page is modified.
.SH NAME
arc \- the fully extensible, open and scriptable build tool
.SH SYNOPSIS
.B arc
[
.B options
]
.I statement
.\"
.SH DESCRIPTION
.I Arc
is a utility to maintain source trees of programs.  It is used to build
programs, libraries, documentations, and maintain dependecies of 
source files.  \fIArc\fP is used to process \fBArcfiles\fP.

.I Arc
is a complete replacement for the make utility, but offers large
scripting possibilities and is platform transparent.

.SH OPTIONS
.TP 5
.BI \-C " dir"
Change to directory \fIdir\fP before reading a build script or doing
anything else.  The 'current working directory' is moved to \fIdir\fP,
so that configuration files, etc. are located in the given directory,
not in the current working directory, from where arc has been called.
.TP 5
.BI \-f " file"
Load \fIfile\fP as build script instead of utilizing the normal search
for a build file.  If the option \fI-r\fP is specified, too, search's for
this file recursively upwards the directory hierachy.
.TP 5
.B \-r
By default arc searchs for a build script only in the current working
directory.  If the \fI-r\fP options is specified it tries to search for a
build script recursively upwards from the current working directory and
uses the first found.
.TP 5
.B \-k
Continue as much as possible after an error; continue if statement failed
.TP 5
.BI \-o " os"
Assumes the operating system \fIos\fP during evaluation.  \fIOS\fP is set by default to
the symbolic name of the current machine but using this switch one can
force the selection of targets with special \fIos:\fP properties
.TP 5
.B \-i
Show information about the callable statements in the script to be loaded
and exit
.TP 5
.B \-v
Be verbose.
.TP 5
.B \-h
Print an informative help message describing the options and then exit.
.TP 5
.B \-V
Print the version number of Arc on the standard error output and then exit.

.SH "SEE ALSO"
The full documentation for
.B arc
is maintained as a Texinfo manual.  If the
.B info
and
.B arc
programs are properly installed at your site, the command
.IP
.B info arc
.PP
should give you access to the complete manual.

.BR make "(1)"
.br
.SH AUTHOR
Gregor Klinke <gck@@eyestep.org>.


ChangeLog
2009-03-20 18:36  gnubryant

	* Arcfile, COPYING, ChangeLog, arc.config.in,
	  boot/bootstrap-asc.scm, boot/bootstrap-guile.scm,
	  boot/bootstrap-ksi.scm, boot/bootstrap-scm.scm,
	  boot/bootstrap.scm, boot/bstr-beos.scm, boot/bstr-unix.scm,
	  boot/bstr-win32.scm, doc/arc.1.in, doc/copyright.scm,
	  src/arc-asc.scm, src/arc-guile.scm, src/arc-ksi.scm,
	  src/arc-scm.scm, src/arc.scm, src/arcconf.scm, src/attrval.scm,
	  src/bt-tasks.scm, src/config.scm.in, src/ctx.scm, src/deps.scm,
	  src/eval.scm, src/excp.scm, src/filter.scm, src/fnmatch.scm,
	  src/getopt.scm, src/handler-factory.scm, src/hash.scm,
	  src/implicit.scm, src/logical.scm, src/meta.scm, src/misc.scm,
	  src/oop.scm, src/path.scm, src/pregexp.scm, src/require.scm,
	  src/show.scm, src/strings.scm, src/sys-beos.scm, src/sys-bsd.scm,
	  src/sys-cygwin.scm, src/sys-darwin.scm, src/sys-linux.scm,
	  src/sys-sunos.scm, src/sys-unix.scm, src/sys.scm, src/sysnm.scm,
	  src/task-util.scm, src/traverse.scm, src/version.scm.in,
	  src/meta-inf/beos.decl, src/meta-inf/bsd.decl,
	  src/meta-inf/cygwin.decl, src/meta-inf/darwin.decl,
	  src/meta-inf/linux.decl, src/meta-inf/sunos.decl,
	  src/tasks/task-bison.scm, src/tasks/task-c-compile.scm,
	  src/tasks/task-c-deps.scm, src/tasks/task-clean.scm,
	  src/tasks/task-copy.scm, src/tasks/task-cvs.scm,
	  src/tasks/task-delete.scm, src/tasks/task-dist.scm,
	  src/tasks/task-file-type.scm, src/tasks/task-fileset.scm,
	  src/tasks/task-filter.scm, src/tasks/task-flex.scm,
	  src/tasks/task-gunzip.scm, src/tasks/task-gzip.scm,
	  src/tasks/task-install.scm, src/tasks/task-java-compile.scm,
	  src/tasks/task-lib.scm, src/tasks/task-link.scm,
	  src/tasks/task-mkdir.scm, src/tasks/task-objc-compile.scm,
	  src/tasks/task-orange-compile.scm,
	  src/tasks/task-orange-deps.scm, src/tasks/task-tar.scm,
	  src/tasks/task-texinfo.scm, src/tasks/task-touch.scm,
	  src/tasks/task-untar.scm, src/tasks/task-version.scm,
	  src/tasks/beos/task-c-compile-beos.scm,
	  src/tasks/beos/task-lib-beos.scm,
	  src/tasks/cygwin/task-c-compile-cygwin.scm,
	  src/tasks/cygwin/task-lib-cygwin.scm,
	  src/tasks/cygwin/task-link-cygwin.scm,
	  src/tasks/darwin/task-c-compile-darwin.scm,
	  src/tasks/darwin/task-lib-darwin.scm,
	  src/tasks/darwin/task-link-darwin.scm,
	  src/tasks/generic/task-c-compile-gcc.scm,
	  src/tasks/generic/task-c-compile-generic.scm,
	  src/tasks/generic/task-lib-generic.scm,
	  src/tasks/generic/task-link-generic.scm,
	  src/tasks/linux/task-c-compile-linux.scm,
	  src/tasks/linux/task-lib-linux.scm,
	  src/tasks/linux/task-link-linux.scm,
	  src/tasks/sunos/task-lib-sunos.scm,
	  src/tasks/sunos/task-link-sunos.scm, test/Arcfile: Updating
	  license to LGPLv3; updating to less CVS-centric code.

2009-03-20 16:45  gnubryant

	* doc/: arc.texinfo.in, fdl.texinfo: Title missing the @unnumbered
	  next to title - version from FSF @c.

2009-03-20 16:32  gnubryant

	* Arcfile, INSTALL, asc/scmfig.h, asc/time.c: Updated to reflect
	  fdl.texinfo add and gpl.texinfo removal.

2009-03-20 16:21  gnubryant

	* doc/arc.texinfo.in: Minor paragraph adjustments, update to the
	  license information in the ignore section.

2009-03-20 15:47  gnubryant

	* doc/: arc.texinfo.in, fdl.texinfo, gpl.texinfo: Updated the Arc
	  manual from GPLv2 to FDLv1.3.  Removed the old gpl.texinfo file,
	  added the new fdl.texinfo file.

2009-03-17 18:48  gnubryant

	* doc/arc.texinfo.in: Performed basic spell checking (manually).
	  Built and tested.  First in a series of technical edits that will
	  include US English spelling, grammar, and readability updates.

2009-03-15 07:30  eyestep

	* ChangeLog: Updated changelog.

2009-03-14 16:40  eyestep

	* Arcfile, asc/findexec.c, asc/sys.c, doc/arc.texinfo.in,
	  src/ctx.scm, src/meta.scm, src/show.scm,
	  src/meta-inf/darwin.decl,
	  src/tasks/darwin/task-c-compile-darwin.scm,
	  src/tasks/darwin/task-lib-darwin.scm,
	  src/tasks/darwin/task-link-darwin.scm,
	  src/tasks/generic/task-lib-generic.scm,
	  src/tasks/linux/task-c-compile-linux.scm: More on Mac OSX support
	  - Fixed a two compiler warning/errors of asc on Mac OSX - Added c
	  compile, lib, and link support for Mac OSX - os: attribute on
	  statement accepts now a list of os names also Asc can not be
	  build with arc on Mac OSX, too

2009-03-14 09:20  eyestep

	* configure, asc/scm.c, boot/bootstrap.scm, src/misc.scm,
	  src/sys-darwin.scm, src/sys.scm, src/sysnm.scm: Added basic
	  support for darwin (osx)

2003-05-05 14:32  eyestep

	* src/tasks/task-bison.scm: added verbose? option

2003-05-05 14:32  eyestep

	* ChangeLog: updated

2003-04-24 17:06  eyestep

	* INSTALL: updated

2003-04-24 17:05  eyestep

	* configure: made generated scripts more portable on solaris shells

2003-04-24 17:00  eyestep

	* boot/bstr-unix.scm: made generated scripts more portable on
	  solaris shells

2003-04-24 16:57  eyestep

	* src/: meta-inf/sunos.decl, tasks/sunos/task-lib-sunos.scm,
	  tasks/sunos/task-link-sunos.scm: added

2003-04-24 16:05  eyestep

	* ChangeLog: updated

2003-04-24 15:59  eyestep

	* INSTALL, README, test/Arcfile, test/version.h.in: updated

2003-04-24 15:49  eyestep

	* asc/sys.c: removed the tmpnam function

2003-04-24 15:49  eyestep

	* Arcfile: dumbed version

2003-04-24 15:48  eyestep

	* configure: fixed problem with "not found" errors in certain
	  shells

2003-04-24 15:48  eyestep

	* boot/bootstrap.scm: added missing debug var

2003-04-24 15:48  eyestep

	* src/meta-inf/bsd.decl: fixed bad reference

2003-04-24 15:47  eyestep

	* src/tasks/task-link.scm: added docs

2003-04-24 15:46  eyestep

	* doc/arc.texinfo.in: updated

2003-04-24 14:28  eyestep

	* src/: config.scm, version.scm: removed, since generated

2003-04-23 11:49  eyestep

	* test/version.h.in: updated

2003-04-23 11:48  eyestep

	* src/tasks/: task-c-compile.scm,
	  generic/task-c-compile-generic.scm: support for platforms where
	  shared builds are not different

2003-04-23 11:48  eyestep

	* src/deps.scm: mtime fixed


src/arc.scm
;;  This file is part of the arc package
;;  Copyright (C) 2002, 2003, 2009 by Gregor Klinke
;;
;;  This library is free software: you can redistribute it and/or modify
;;  it under the terms of the GNU Lesser General Public License as published
;;  by the Free Software Foundation, either version 3 of the License, or
;;  (at your option) any later version.
;;
;;  This library is distributed in the hope that it will be useful,
;;  but WITHOUT ANY WARRANTY; without even the implied warranty of
;;  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;;  GNU Lesser General Public License for more details.
;;
;;  You should have received a copy of the GNU General Public License
;;  along with this program.  If not, see <http://www.gnu.org/licenses/>.

;; setup the require/provide system
(define %arc:debug% (let ((ff (getenv "ARC_DEBUG")))
                      (if (and ff (equal? ff "yes")) #t #f)))

;; define global variables and flags
(define %arc:verbose% #f)

(define arc:script-name #f)
(define arc:config-name #f)
(define arc:current-stmt #f)
(define %arc:find-script-rec% #f)
(define %arc:start-dir% ())
(define %arc:keep-going-on-errors% #f)
(define %arc:show-infos-only% #f)

(define arc:default-script-names '("ARCFile" "Arcfile" "ArcFile" "arcfile"
				   "ARCFILE" "build.arc" "Build.arc" 
				   "build.ARC" "Build.ARC" "BUILD.ARC"))
(define arc:default-config-names '("arc.config"
                                   "ARCConfig" "Arcconfig" "ArcConfig" 
                                   "arcconfig" "ARCCONFIG" "Config.arc" 
                                   "config.arc" "Arc.rc" "arc.rc"))

(load (string-append %arc:home% "/config.scm"))
(load (string-append %arc:home% "/misc.scm"))
(load (string-append %arc:home% "/sysnm.scm"))
(load (string-append %arc:home% "/strings.scm"))
(load (string-append %arc:home% "/logical.scm"))
(load (string-append %arc:home% "/excp.scm"))

(define %arc:sysnm% (arc:canonical-sysnm (arc:host-os)
                                         (arc:host-maker)
                                         (arc:host-cpu)
                                         (arc:host-version)))

(load (string-append %arc:home% "/oop.scm"))
(load (string-append %arc:home% "/sys.scm"))



(if (arc:sys 'file-exists? (string-append %arc:home% "/version.scm"))
    (load (string-append %arc:home% "/version.scm"))
    (define %arc:version% "?"))


;; this variable defines the os as used for the evaluation machine.  This
;; does not need to be identical to the real system name
(define %arc:eval-os% (car %arc:sysnm%))

;; ----------------------------------------------------------------------
;; set up the include path
;; ----------------------------------------------------------------------
(define %arc:arc-incl-path% 
  (let ((ip (or (getenv "ARC_INCL_PATH")
                (arc:include-path))))
    (if ip
        (append (append (arc:split-string ip (arc:pathlist-sep))
                        (list %arc:home%))
                (list (string-append %arc:home% "/tasks")))
        (append (list %arc:home%)
                (list (string-append %arc:home% "/tasks"))))))


;; ----------------------------------------------------------------------
;; set up the system environment
;; ----------------------------------------------------------------------
(if %arc:debug%
    (arc:display "scheme interpreter: " %arc:scheme-impl% #\nl
                 "system name:        " %arc:sysnm% #\nl
                 "arc home:           " %arc:home% #\nl))

(define %arc:include-list%
  '("require.scm"
    "getopt.scm"
    "path.scm"
    "hash.scm"
    "eval.scm"
    "arcconf.scm"
    "show.scm"
    "implicit.scm"
    "ctx.scm"
    "meta.scm"
    "deps.scm"
    "task-util.scm"
    "traverse.scm"
    "attrval.scm"
    "filter.scm"
    "pregexp.scm"
    "fnmatch.scm"
    "handler-factory.scm"
    "bt-tasks.scm"))
  
(let loop ((fn %arc:include-list%))
  (if (null? fn)
      #t
      (begin
        (load (string-append %arc:home% "/" (car fn)))
        (loop (cdr fn)))))


;; display a help text
(define (arc:display-help)
  (arc:display
   "arc - a scheme based makefile and config system" #\nl
   "usage:" #\nl
   "  arc [options] [statement]" #\nl
   #\nl
   "Options:" #\nl
   " -f FILE   Load FILE as build script instead of 'Arcfile' as default" #\nl
   " -C DIR    Change to directory DIR before reading a build script or doing" #\nl
   "           anything else" #\nl
   " -r        By default arc looks for the build script 'build.arc' in the" #\nl
   "           current working directory.  If the -r options is specified it" #\nl
   "           tries to search for a build script recursively upwards from the" #\nl
   "           current working directory an uses the first found." #\nl
   " -k        Continue as much as possible after an error; continue if " #\nl
   "           statement failed" #\nl
   " -v        be verbose" #\nl
   " -o OS     assume the operating system OS during evaluation.  OS is" #\nl
   "           set by default to the symbolic name of the current machine," #\nl
   "           but using this switch one can force the selection of targets" #\nl
   "           with special os: properties" #\nl
   " -i        show information about the callable statements in the" #\nl
   "           script to be loaded and exit" #\nl
   " -h        This help" #\nl
   " -V        Print the version and exit" #\nl
   #\nl 
   "if no statement is given on the command line arc uses the default statement " #\nl
   "as stated in the defproject statement in the build script used." #\nl  
   #\nl #\nl))

(define (arc:display-version)
  (arc:display
   "Arc version " %arc:version% " (" %arc:scheme-impl% ")" #\nl
   "Copyright (C) 2002, 2003 Gregor Klinke" #\nl
   "This is free software; see the source for copying conditions." #\nl
   "There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A" #\nl
   "PARTICULAR PURPOSE." #\nl))


(define arc:opts '((verbose   "-v" "--verbose" #f)
                   (script    "-f" "--script" #t)
                   (conf      "-c" "--conf" #t)
                   (recursive "-r" "--rec" #f)
                   (keep      "-k" "--keep-going" #f)
                   (change    "-C" "--dir" #t)
                   (help      "-h" "--help" #f)
                   (info      "-i" "--info" #f)
                   (os        "-o" "--os" #t)
                   (version   "-V" "--version" #f)))
(let loop ((opt (arc:getopt %arc:argv% arc:opts)))
  (if (not opt)
      #f
      (let ((done #f))
        (case opt
          ((verbose) (set! %arc:verbose% #t))
          ((conf) (set! arc:config-name *arc:optarg*))
          ((script) (set! arc:script-name *arc:optarg*))
          ((recursive) (set! %arc:find-script-rec% #t))
          ((keep) (set! %arc:keep-going-on-errors% #t))
          ((os) (set! %arc:eval-os% (string->symbol *arc:optarg*)))
          ((change) (begin
                      ;; store the current directory
                      (set! %arc:start-dir% (arc:path-cwd))
                      ;; change to new working directory
                      (arc:sys 'chdir *arc:optarg*)))
          ((help) (begin
                    (arc:display-help)
                    (quit)))
          ((version) (begin
                       (arc:display-version)
                       (quit)))
          ((info) (set! %arc:show-infos-only% #t))
          ((#\?) (arc:msg "bad option: " *arc:optopt*))
          ((#\:) (begin
                   (arc:msg "missing arg " *arc:optopt*)))
          (else (begin 
                  (set! arc:current-stmt (string->symbol opt))
                  (set! done #t))))
        (if (not done)
            (loop (arc:getopt %arc:argv% arc:opts)))) ))


(if %arc:verbose%
    (begin
      (arc:display "arc system version " %arc:version% #\nl)
      (arc:display "assume os: " %arc:eval-os% #\nl)))


;; start the scripting logic
;; [1] look for a arcconfig file to load
(let ((script (if arc:config-name 
		  arc:config-name
		  (let loop ((nl arc:default-config-names))
		    (if (null? nl)
			#f
			(or (arc:find-script (car nl))
			    (loop (cdr nl))))))))
  (if %arc:verbose%
      (arc:msg "use config file: " script))
  (if script
      (arc:load-arcconfig script)))


;; [2] look for the Arcfile and go ...
(let ((script (if arc:script-name 
		  arc:script-name
		  (let loop ((nl arc:default-script-names))
		    (if (null? nl)
			#f
			(or (arc:find-script (car nl))
			    (loop (cdr nl))))))))
  (if %arc:verbose%
      (arc:msg "use script: " script))
  (if script
      (if %arc:show-infos-only%
          (arc:display-script script)
          (arc:evaluate-script script 'public))
      (arc:msg "no script file found")))


;; finish
(quit)


;;Keep this comment at the end of the file 
;;Local variables:
;;mode: scheme
;;End:
